idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/references/FirReferenceResolveHelper.kt

```diff
@@ -374,42 +375,104 @@ internal object FirReferenceResolveHelper {
         fir: FirResolvedQualifier,
         expression: KtSimpleNameExpression,
         session: FirSession,
-        symbolBuilder: KtSymbolByFirBuilder,
-        analysisSession: KtFirAnalysisSession
+        symbolBuilder: KtSymbolByFirBuilder
     ): Collection<KtSymbol> {
-        // TODO refactor that block
-        val classId = fir.classId ?: return listOfNotNull(getPackageSymbolFor(expression, symbolBuilder, forQualifiedType = false))
+        val referencedSymbol = if (fir.resolveToCompanionObject) {
+            (fir.symbol?.fir as? FirRegularClass)?.companionObject?.symbol
+        } else {
+            fir.symbol
+        }
+        if (referencedSymbol == null) {
+            // If referencedSymbol is null, it means the reference goes to a package.
+            val parent = expression.parent as? KtDotQualifiedExpression ?: return emptyList()
+            val fqNameSegments =
+                when (expression) {
+                    parent.selectorExpression -> parent.fqNameSegments() ?: return emptyList()
+                    parent.receiverExpression -> listOf(expression.getReferencedName())
+                    else -> return emptyList()
+                }
+            return listOfNotNull(symbolBuilder.createPackageSymbolIfOneExists(FqName.fromSegments(fqNameSegments)))
+        }
+        val referencedClass = referencedSymbol.fir
+        val referencedSymbolsByFir = listOfNotNull(symbolBuilder.buildSymbol(referencedClass))
+        if (fir.source.psi === expression) return referencedSymbolsByFir
```

Comment:
A little confused by this. When is this condition true, when is it false?

Reply:
The source PSI of a `FirResolvedTypeRef` is a `KtSimpleNameExpression` if it's not qualified. For example
```
object A {
  fun foo() {}
}
fun test() {
  A.foo() // the above condition is true for `A`.
}
```

Reply:
Ok. A comment would help so we know that for the rest of the function, the PSI is a qualified expression. Is it expected that the PSI would be a `KtDotQualifiedExpression` from here on? We can add an assertion too.

I assume you meant `FirResolvedQualifier` in your reply?